blockchain/src/blockchain/verify.rs

```diff
@@ -311,71 +310,45 @@ impl Blockchain {
-        signed_data: Vec<u8>,
-        signature: &Signature,
-        signature_only: bool,
     ) -> Result<MacroBody, PushError> {
         // If the accounts tree is not complete nothing can be done
         if !self.accounts_complete() {
             // Maybe introduce a proper error here. Is this even needed?
             return Err(PushError::IncompleteAccountsTrie);
         }
 
-        // No inherent was given, but signature verification is the only required step, independent from
-        // the result this can never work, as the inherent would have to be returned.
-        if proposed_block.body.is_none() && signature_only {
-            return Err(PushError::InvalidBlock(BlockError::MissingBody));
-        }
-
         // Get the seed of the preceding block.
         let prev_header = self
             .get_block(&proposed_block.header.parent_hash, false, None)
             .map_err(PushError::BlockchainError)?
             .header();
 
-        // Create a read transaction to use in the following queries.
-        // It is not necessary for the correct execution and only is created to not create 3 different
-        // ones within the 3 function calls making use of it here.
-        let txn = self.read_transaction();
-
-        // Get the validator for this round.
-        let mut proposer = self
-            .get_proposer_at(
-                proposed_block.block_number(),
-                round,
-                prev_header.seed().entropy(),
-                Some(&txn),
-            )
-            .expect("Couldn't find slot owner!")
-            .validator
-            .signing_key;
-
-        // Verify the signature. The proposal is signed by the proposer of the round the proposal is used in.
-        if !proposer.verify(signature, signed_data.as_slice()) {
-            return Err(PushError::InvalidBlock(BlockError::InvalidJustification));
-        }
-
-        if signature_only {
-            return Ok(proposed_block.body.unwrap());
-        }
-
         // The VRF seed will be signed by the proposer of the original round the proposal was proposed in.
-        // The round is specified within the header itself, and it should only ever differ from the `round`
-        // value provided if `valid_round` is Some(vr). Importantly though, that very vr is not necessarily
+        // The round is specified within the header itself, and it should only ever differ from the round
+        // value provided if valid_round = Some(vr). Importantly though, that very vr is not necessarily
         // the same as the round given in the header, as it might not be the first time this proposal is
         // re-proposed.
-        if proposed_block.header.round != round {
+        if let Some(vr) = valid_round {
+            if vr < proposed_block.header.round || vr >= round {
+                // A VR only makes sense if it is at least the original round of the proposal,
+                // while also not being in the future.
```

Comment:
Yes it can and usually will be. 

For any proposal that is re-proposed the first time it will have VR being the round of the original proposal. Only once this iteration gets re-proposed it will have as the VR the round it was re-proposed in.